src/time.cpp

```diff
@@ -87,20 +87,27 @@ namespace g3 {
 
    int timespec_get(struct timespec* ts/*, int base*/) {
 #ifdef __MACH__
-      // std::timespec_get or posix clock_gettime)(...) are not
-      // implemented on OSX
-      // @return value of base if successful, else zero
-      struct timeval now = {};
-      int rv = gettimeofday(&now, nullptr);
-      if (-1 == rv) {
-         return rv;
-      }
-      // error mode. just return sec, microsecond
-      ts->tv_sec  = now.tv_sec;
-      ts->tv_nsec = now.tv_usec * 1000;
-      return 0;
+	   // std::timespec_get or posix clock_gettime)(...) are not
+	   // implemented on OSX
+	   // @return value of base if successful, else zero
+	   struct timeval now = {};
+	   int rv = gettimeofday(&now, nullptr);
+	   if (-1 == rv) {
+		   return rv;
+	   }
+	   // error mode. just return sec, microsecond
+	   ts->tv_sec = now.tv_sec;
+	   ts->tv_nsec = now.tv_usec * 1000;
+	   return 0;
 #elif(defined(WIN32) || defined(_WIN32) || defined(__WIN32__))
-	   return timespec_get(ts, TIME_UTC);
+	   static auto os =
+		   std::chrono::time_point_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now()).time_since_epoch() -
+		   std::chrono::time_point_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now()).time_since_epoch();
+
+	   auto now_ns = (std::chrono::time_point_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now()).time_since_epoch() + os).count();
+	   ts ->tv_sec = now_ns / 1000000000;
+	   ts ->tv_nsec = now_ns % 1000000000;
```

Comment:
It calculates the remainder of a division by 1e9 (which in turn yields the full seconds). Most compilers will make a single integer division with remainder out of the two statements (thus I did use the second divison, which % effectively is instead of subtraction tv_sec * 1e9.